modules/core/src/lifecycle/component.js

```diff
@@ -32,7 +32,20 @@ export default class Component {
 
   // clone this layer with modified props
   clone(newProps) {
-    return new this.constructor(Object.assign({}, this.props, newProps));
+    const {props} = this;
+
+    // Async props cannot be copied with Object.assign. Extract the plain value.
+    const asyncProps = {};
+
+    for (const key in props._asyncPropDefaultValues) {
+      if (key in props._asyncPropResolvedValues) {
+        asyncProps[key] = props._asyncPropResolvedValues[key];
+      } else if (key in props._asyncPropOriginalValues) {
+        asyncProps[key] = props._asyncPropOriginalValues[key];
+      }
+    }
+
+    return new this.constructor(Object.assign({}, props, asyncProps, newProps));
```

Comment:
Nit: Not part of the scope of the change in this PR, feel free to ignore:

When I read this, I suspect that `clone` is not participating in the optimization that avoids/minimizes the repeated copying of all props using `Object.assign`?

Would ` return new this.constructor(props, asyncProps, newProps);` work instead? I think that would avoid one round of `Object.assign`?

Maybe add a bench case for clone together with the existing layer instantiation cases and then do that change? At least so we know if `clone` has a similar perf as new.

Reply:
Unfortunately, a bunch of core layers ([example](https://github.com/uber/deck.gl/blob/master/modules/layers/src/line-layer/line-layer.js#L41)) do not support constructor with multiple arguments, which is another bug I'll fix. Potentially any custom layers created by the user may override `constructor` incorrectly too.

While this solution might be slower, it is guaranteed to clone a layer safely.